src/services/mapCode.ts

```diff
@@ -0,0 +1,365 @@
+    if (!focusLocations || !focusLocations.length) {
+        changeTracker.insertNodesAtEndOfFile(file, nodes, /*blankLineBetween*/ false);
+        return;
+    }
+
+    // Otherwise, we'll need to find the right place to insert or replace.
+
+    // First, we try and find a nearby frame of reference: are there
+    // declarations with similar names nearby?
+    for (const locationGroup of focusLocations) {
+        for (const location of locationGroup) {
+            const scope: Node | undefined = findAncestor(getNodeAtPosition(file, location.start), node => {
+                if (!isBlock(node) && !isSourceFile(node)) {
+                    return false;
+                }
+                for (const stmt of node.statements) {
+                    for (const node of nodes) {
+                        if (matchNode(node, stmt)) {
+                            return true;
+                        }
+                    }
+                }
+                return false;
+            });
+            if (scope) {
+                // We found a scope that contains a matching node.
+
+                // First we find the last match of the range we're replacing.
+                // That is, assuming we have the following code to start with:
+                //
+                // ```ts
+                // function foo() {
+                //     console.log("hello");
+                // }
+                // function bar() {
+                //     console.log("world");
+                // }
+                // function baz() {
+                //     console.log("!");
+                // }
+                // function other() {
+                //    console.log("Leave me alone!");
+                // }
+                // ```
+                //
+                // And we get a map request with the following code:
+                //
+                // ```ts
+                // function bar() {
+                //     console.log("world");
+                // }
+                // function baz() {
+                //     console.log("!");
+                // }
+                // function qux() {
+                //     console.log("?");
+                // }
+                // ```
+                //
+                // `bar` will be the beginning of our range, and we want to
+                // have `baz` in the original be the end of our replacement
+                // range. `other` should be left intact, while `qux` should be
+                // "injected" before it.
+                //
+                // We do this by scanning the scope for all nodes that have a
+                // match in the incoming map request code, then sorting and
+                // picking first and last by position, and that will be our
+                // range for replacement.
+                const matches = (scope as Block).statements.filter(stmt => nodes.some(node => matchNode(node, stmt)));
+                matches.sort((a, b) => a.pos - b.pos);

```

Comment:
They won't be if, say, Copilot reordered two functions. At least if I'm thinking about it right.

Reply:
Since the code deletes all of the original matched statements, the order that any corresponding statements generate in copilot shouldn't matter since the copilot content just gets "pasted" in. I also had to think about this a couple times.

Instead of filtering, since we only use the first and last match, we could just `find()` and `findLast()` through `scope.statements`?